common/transport/amqp/src/amqp.ts

```diff
@@ -135,6 +88,280 @@ export class Amqp {
+            };
+
+            this._receivers[endpoint].on('error', attachErrorHandler);
+            this._receivers[endpoint].attach((err) => {
+              if (err) {
+                debug('failed to attach receiver link: ' + endpoint + ': ' + err.toString());
+                done(err);
+              } else {
+                this._receivers[endpoint].removeListener('error', attachErrorHandler);
+                this._receivers[endpoint].on('error', (err) => {
+                  debug('error on sender link: ' + endpoint + ': ' + err.toString());
+                  delete(this._receivers[endpoint]);
+                });
+                debug('sender link attached: ' + endpoint);
+                done(null, this._receivers[endpoint]);
+              }
+            });
+          },
+          attachSenderLink: (endpoint: string, linkOptions: any, done: GenericAmqpBaseCallback<any>): void => {
+            debug('creating sender link for: ' + endpoint);
+            let senderFsm = new SenderLink(endpoint, linkOptions, this._amqp);
+            this._senders[endpoint] = senderFsm;
+            const attachErrorHandler = (err) => {
+              debug('sender link error event while attaching: ' + endpoint + ': ' + err.toString());
+              delete(this._senders[endpoint]);
+              done(err);
+            };
+
+            this._senders[endpoint].on('error', attachErrorHandler);
+            debug('attaching sender link for: ' + endpoint);
+            this._senders[endpoint].attach((err) => {
+              if (err) {
+                debug('failed to attach sender link: ' + endpoint + ': ' + err.toString());
+                done(err);
+              } else {
+                this._senders[endpoint].removeListener('error', attachErrorHandler);
+                this._senders[endpoint].on('error', (err) => {
+                  debug('error on sender link: ' + endpoint + ': ' + err.toString());
+                  delete(this._senders[endpoint]);
+                });
+                debug('sender link attached: ' + endpoint);
+                done(null, this._senders[endpoint]);
+              }
+            });
+          },
+          detachReceiverLink: (endpoint: string, detachCallback: GenericAmqpBaseCallback<any>): void => {
+            if (!this._receivers[endpoint]) {
+              this._safeCallback(detachCallback);
+            } else {
+              this._detachLink(this._receivers[endpoint], (err?) => {
+                delete(this._receivers[endpoint]);
+                this._safeCallback(detachCallback, err);
+              });
+            }
+          },
+          detachSenderLink: (endpoint: string, detachCallback: GenericAmqpBaseCallback<any>): void => {
+            this._detachLink(this._senders[endpoint], (err?) => {
+              delete(this._senders[endpoint]);
+              this._safeCallback(detachCallback, err);
+            });
+          }
+        },
+        disconnecting: {
```

Comment:
>disconnecting [](start = 8, length = 13)

This block is difficult to read.  2 things slowed me down:
1) names disconnectFunc and detachFunc are confusing.  I knew disconnectFunc has a function that had something to do with disconnect, but I didn't know if it was a callback or the function that actually disconnected.  Suggest naming detachFunc to detachLink following the verb-subject pattern.  Same for disconnectFunc -- disconnectConnection doesn't sound right, but I don't know what is better.
2) The ternary operators you used in disconnectFunc and detachFunc are unusual.  I would expect the comparison to happen inside the function -- this is more typical and easier (for me) to grok.